dropshot/src/config.rs

```diff
@@ -113,9 +116,108 @@ impl Default for ConfigDropshot {
     fn default() -> Self {
         ConfigDropshot {
             bind_address: "127.0.0.1:0".parse().unwrap(),
-            request_body_max_bytes: 1024,
+            default_request_body_max_bytes: 1024,
             default_handler_task_mode: HandlerTaskMode::Detached,
             log_headers: Default::default(),
         }
     }
 }
+
+#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
+#[serde(default)]
+struct DeserializedConfigDropshot {
+    bind_address: SocketAddr,
+    default_request_body_max_bytes: usize,
+    // Previous name for default_request_body_max_bytes. Present only to guide
+    // users to the new name.
+    #[serde(
+        deserialize_with = "deserialize_invalid_request_body_max_bytes",
+        skip_serializing_if = "Option::is_none"
+    )]
+    request_body_max_bytes: Option<InvalidConfig>,
+    default_handler_task_mode: HandlerTaskMode,
+    log_headers: Vec<String>,
+}
+
+impl From<DeserializedConfigDropshot> for ConfigDropshot {
+    fn from(v: DeserializedConfigDropshot) -> Self {
+        ConfigDropshot {
+            bind_address: v.bind_address,
+            default_request_body_max_bytes: v.default_request_body_max_bytes,
+            default_handler_task_mode: v.default_handler_task_mode,
+            log_headers: v.log_headers,
+        }
+    }
+}
+
+impl From<ConfigDropshot> for DeserializedConfigDropshot {
+    fn from(v: ConfigDropshot) -> Self {
+        DeserializedConfigDropshot {
+            bind_address: v.bind_address,
+            default_request_body_max_bytes: v.default_request_body_max_bytes,
+            request_body_max_bytes: None,
+            default_handler_task_mode: v.default_handler_task_mode,
+            log_headers: v.log_headers,
+        }
+    }
+}
+
+impl Default for DeserializedConfigDropshot {
+    fn default() -> Self {
+        ConfigDropshot::default().into()
+    }
+}
+
+/// A marker type to indicate that the configuration is invalid.
+///
+/// This type can never be constructed, which means that for any valid config,
+/// `Option<InvalidConfig>` is always none.
+#[derive(Clone, Debug, PartialEq, Serialize)]
+pub enum InvalidConfig {}
+
+fn deserialize_invalid_request_body_max_bytes<'de, D>(
```

Comment:
do we prefer this to `impl Serialize for InvalidConfig`?

Reply:
This pattern allows a custom error message to be passed in -- `InvalidConfig` is a general marker value that shouldn't necessarily be tied with the specific error message.

I'll add a comment to this note.

Reply:
Oh also this returns an `Option<InvalidConfig>` which wouldn't be possible to do this way.

Reply:
> Oh also this returns an `Option<InvalidConfig>` which wouldn't be possible to do via a type implementation.

that's true... but I think `impl Serialize for InvalidConfig` would probably suffice since the implementation of Serialize for Option will call it.

Reply:
Hmm, `InvalidConfig` doesn't implement `Serialize` at all (the `skip_serializing` makes that moot). This code only does deserialization.